### PR TITLE
gpconfig: add validation for readonly GUCs

### PR DESCRIPTION
**Pull request by larham**
-- metadata about gucs, like GUC_DISALLOW_IN_FILE, is not available via
sql. Work around that by parsing the C code that defines this metadata, and store
relevant information in a file, to be read by gpconfig when called to
change a GUC.

-- the Makefile in gpMgmt/bin will create a local file during 'make install', used by
gpconfig to validate that a given GUC can be changed

Signed-off-by: C.J. Jameson <cjameson@pivotal.io>

### CODE REVIEW

**Comment by hlinnaka** [comment]
It's really hard to follow how this works, by reading the code. Needs more comments, I guess.

But overall, I it would be easier and more robust to add a backend function to return the information that you need.

**Comment by skahler-pivotal** [comment]
Being able to pull that information via a function would be nice to have, though I don't think it will give us what we would need in the end here.

For gpconfig I'd like to be able to know what they are without having to have the database running. At this point in time one of the issues with gpconfig is the database needs to be running in order to run it to pull segment info. We are going to do things to try to back away from this restriction. I'd like to avoid it having to need to have the database running to get this additional pieces of information.

So I'd like to see us referencing a file rather than a function in gpconfig.

File: gpMgmt/bin/gpconfig [hunk 1]
```diff
@@ -29,6 +30,7 @@ EXECNAME = os.path.split(__file__)[-1]
 
 PROHIBITED_GUCS = set(["port", "listen_addresses"])
 SAMEVALUE_GUCS = set(["gp_default_storage_options"])
+read_only_gucs = set([])
```
Comment (Chibin) [inline gpMgmt/bin/gpconfig]: Just clarifying, this is not all caps because we're changing the values in the variable?

Comment (larham) [inline gpMgmt/bin/gpconfig]: Yes, instead of a static set, this set of read_only_gucs is populated at runtime, from the set of all GUCS defined by guc.c and guc_gp.c.  In that sense, it is not a static list known ahead of time by this python code, even though that list of read_only_gucs is defined in C as a static list. The language barrier makes a runtime transfer of information necessary.

File: gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import os
+import re
+
+# this finds all instances where a property name that is both surrounded with quotes and is adjacent
+# to the metadata "GUC_DISALLOWED_IN_FILE", where adjacent means not separated by a closing } bracket
```
Comment (Chibin) [inline gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py]: Biggest assumption about this regex is that we're turning the whole file into a single string.

Comment (Chibin) [inline gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py]: s/GUC_DISALLOWED_IN_FILE/GUC_DISALLOW_IN_FILE

File: gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import os
+import re
+
+# this finds all instances where a property name that is both surrounded with quotes and is adjacent
```
Comment (Chibin) [inline gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py]: 👍

**Comment by Chibin** [comment]
> But overall, I it would be easier and more robust to add a backend function to return the information that you need.

It'd be great having this, I'm interested in trying to make this work, but it'll probably be out of this PR.

**Comment by cjcjameson** [comment]
Manual review so far this afternoon looks good. I want to do some more testing on the packaged artifact to make sure it gets the file, and we could even write an automated validation into the Concourse packaging step to test-drive the presence of this (one-off and surprising) file.

File: gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import errno
+import os
+import re
+
+# this finds all instances of a property name that is both surrounded with quotes and is adjacent
```
Comment (larham) [inline gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py]: this comment got separated from the pattern by a few lines because of an IDE-added import.  Please move it next to the pattern, with the added benefit of putting the import lines all together.

Comment (Chibin) [inline gpMgmt/bin/gpconfig_modules/parse_guc_metadata.py]: Fixed

**Comment by cjcjameson** [comment]
I set a forked pipeline on this branch to see what the packaged artifact will look like so we can assess how it is to have the new `gucs_disallowed_in_file.txt` baked in.

**Comment by cjcjameson** [comment]
Idea(s) for balancing "doesn't rely on this magic flat file" vs. "available when the database isn't running":

- Would it be common for functions/utilities/programs in `contrib/` to be runnable while the database is down, but also "hook into" the backend of the database somewhat easily? Is that the right layer at which to build a simple guc reader that is somewhat removed from relying on a running database? @hlinnaka @skahler-pivotal 

Given this magic flat file listing read-only GUCs, here are ideas for more protection or assurances: @tushar-dadlani @larham 
- Store a checksum on the flat file and assert that it itself hasn't been tampered with after build / initialization
- Tests in the gpdb_rc_packaging_centos concourse job to assert that the new file is there

**Comment by cjcjameson** [comment]
rebased to get fresh PR pipeline run

**Comment by Chibin** [comment]
> Manually verified that the new text file shows up at share/greenplum/gucs_disallowed_in_file.txt in the compilation process that Concourse runs. @Chibin can you evaluate and triage the above ideas? But otherwise looks good to me.

I've noted your thoughts into our icebox.
